Infinite recursion in SR.GetResourceString when using debug ILCompiler build

Create a console app containing just this line:
```C#
new System.SystemException();
```
and compile it in release mode with a **debug** build of the ILCompiler package:
```cmd
dotnet publish -r win-x64 -c Release
```
If you run it, the app goes into infinite recursion in the `SR.GetResourceString` method, eventually failing due to stack overflow (that takes a long time though):
```
...
Repro!S_P_CoreLib_System_Diagnostics_Debug__Fail_0
Repro!S_P_CoreLib_System_Diagnostics_Debug__Assert_1
Repro!S_P_CoreLib_System_Diagnostics_Debug__Assert
Repro!S_P_TypeLoader_Internal_Runtime_TypeLoader_TypeLoaderEnvironment__TryGetStaticClassConstructionContext
Repro!S_P_Reflection_Execution_Internal_Reflection_Execution_ReflectionExecutionDomainCallbacksImplementation__TryGetStaticClassConstructionContext
Repro!S_P_CoreLib_System_Runtime_CompilerServices_RuntimeHelpers__RunClassConstructor
Repro!S_P_CoreLib_System_SR__InternalGetResourceString
𝗥𝗲𝗽𝗿𝗼!𝗦_𝗣_𝗖𝗼𝗿𝗲𝗟𝗶𝗯_𝗦𝘆𝘀𝘁𝗲𝗺_𝗦𝗥__𝗚𝗲𝘁𝗥𝗲𝘀𝗼𝘂𝗿𝗰𝗲𝗦𝘁𝗿𝗶𝗻𝗴_𝟬
Repro!S_P_CoreLib_System_Diagnostics_StackFrame__AppendToStackTrace
Repro!S_P_CoreLib_System_Diagnostics_StackTrace__ToString
Repro!S_P_CoreLib_System_Diagnostics_StackTrace__ToString_1
Repro!S_P_CoreLib_System_Diagnostics_DebugProvider__Fail
Repro!S_P_CoreLib_System_Diagnostics_Debug__Fail_0
Repro!S_P_CoreLib_System_Diagnostics_Debug__Assert_1
Repro!S_P_CoreLib_System_Diagnostics_Debug__Assert
Repro!S_P_TypeLoader_Internal_Runtime_TypeLoader_TypeLoaderEnvironment__TryGetStaticClassConstructionContext
Repro!S_P_Reflection_Execution_Internal_Reflection_Execution_ReflectionExecutionDomainCallbacksImplementation__TryGetStaticClassConstructionContext
Repro!S_P_CoreLib_System_Runtime_CompilerServices_RuntimeHelpers__RunClassConstructor
Repro!S_P_CoreLib_System_SR__InternalGetResourceString
𝗥𝗲𝗽𝗿𝗼!𝗦_𝗣_𝗖𝗼𝗿𝗲𝗟𝗶𝗯_𝗦𝘆𝘀𝘁𝗲𝗺_𝗦𝗥__𝗚𝗲𝘁𝗥𝗲𝘀𝗼𝘂𝗿𝗰𝗲𝗦𝘁𝗿𝗶𝗻𝗴_𝟬
Repro!S_P_CoreLib_System_SystemException___ctor
Repro!Repro__Module___StartupCodeMain
Repro!wmain
Repro!invoke_main
Repro!__scrt_common_main_seh
Repro!__scrt_common_main
Repro!wmainCRTStartup
KERNEL32!BaseThreadInitThunk
ntdll!RtlUserThreadStart
```
That is caused by the failed assertion in `TypeLoaderEnvironment.TryGetStaticClassConstructionContext` for the `System.Resources.ResourceReader` type: https://github.com/dotnet/runtimelab/blob/4ab2b2bad7f3d1befab60448ed0b4d0a1b3e5c29/src/coreclr/nativeaot/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/TypeLoaderEnvironment.Metadata.cs#L458-L460
To create a debug ILCompiler package, I used the following commands:
```cmd
build -arch x64 -c Release -s nativeaot
build -arch x64 -lc Release -s nativeaot+libs+nativeaot.packages
```
The first command was needed to avoid the following error (which is probably a separate issue):
```
CSC : error CS2001: Source file 'artifacts\obj\coreclr\nativeaot\System.Private.CoreLib\..\..\windows.x64.Release\nativeaot\Runtime\Full\AsmOffsets.cs' could not be found. [src\coreclr\nativeaot\System.Private.CoreLib\src\System.Private.CoreLib.csproj]
```

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/AnalysisBasedMetadataManager.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/AnalysisBasedMetadataManager.cs
@@ -22,6 +22,7 @@ namespace ILCompiler
     public sealed class AnalysisBasedMetadataManager : GeneratingMetadataManager, ICompilationRootProvider
     {
         private readonly List<ModuleDesc> _modulesWithMetadata;
+        private readonly List<MetadataType> _typesWithRootedCctorContext = new List<MetadataType>();
 
         private readonly Dictionary<TypeDesc, MetadataCategory> _reflectableTypes = new Dictionary<TypeDesc, MetadataCategory>();
         private readonly Dictionary<MethodDesc, MetadataCategory> _reflectableMethods = new Dictionary<MethodDesc, MetadataCategory>();
@@ -33,7 +34,8 @@ namespace ILCompiler
                 new FullyBlockedManifestResourceBlockingPolicy(), null, new NoStackTraceEmissionPolicy(),
                 new NoDynamicInvokeThunkGenerationPolicy(), Array.Empty<ModuleDesc>(),
                 Array.Empty<ReflectableEntity<TypeDesc>>(), Array.Empty<ReflectableEntity<MethodDesc>>(),
-                Array.Empty<ReflectableEntity<FieldDesc>>(), Array.Empty<ReflectableCustomAttribute>())
+                Array.Empty<ReflectableEntity<FieldDesc>>(), Array.Empty<ReflectableCustomAttribute>(),
+                Array.Empty<MetadataType>())
         {
         }
 
@@ -48,10 +50,12 @@ namespace ILCompiler
             IEnumerable<ReflectableEntity<TypeDesc>> reflectableTypes,
             IEnumerable<ReflectableEntity<MethodDesc>> reflectableMethods,
             IEnumerable<ReflectableEntity<FieldDesc>> reflectableFields,
-            IEnumerable<ReflectableCustomAttribute> reflectableAttributes)
+            IEnumerable<ReflectableCustomAttribute> reflectableAttributes,
+            IEnumerable<MetadataType> rootedCctorContexts)
             : base(typeSystemContext, blockingPolicy, resourceBlockingPolicy, logFile, stackTracePolicy, invokeThunkGenerationPolicy)
         {
             _modulesWithMetadata = new List<ModuleDesc>(modulesWithMetadata);
+            _typesWithRootedCctorContext = new List<MetadataType>(rootedCctorContexts);
             
             foreach (var refType in reflectableTypes)
             {
@@ -208,6 +212,11 @@ namespace ILCompiler
                             rootProvider.RootNonGCStaticBaseForType(field.OwningType, reason);
                     }
                 }
+            }
+
+            foreach (var type in _typesWithRootedCctorContext)
+            {
+                rootProvider.RootNonGCStaticBaseForType(type, reason);
             }
         }
 

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/AnalysisBasedMetadataManager.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/AnalysisBasedMetadataManager.cs
@@ -22,7 +22,7 @@ namespace ILCompiler
     public sealed class AnalysisBasedMetadataManager : GeneratingMetadataManager, ICompilationRootProvider
     {
         private readonly List<ModuleDesc> _modulesWithMetadata;
-        private readonly List<MetadataType> _typesWithRootedCctorContext = new List<MetadataType>();
+        private readonly List<MetadataType> _typesWithRootedCctorContext;
 
         private readonly Dictionary<TypeDesc, MetadataCategory> _reflectableTypes = new Dictionary<TypeDesc, MetadataCategory>();
         private readonly Dictionary<MethodDesc, MetadataCategory> _reflectableMethods = new Dictionary<MethodDesc, MetadataCategory>();

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs
@@ -794,10 +794,28 @@ namespace ILCompiler
                 }
             }
 
+            var rootedCctorContexts = new List<MetadataType>();
+            foreach (NonGCStaticsNode cctorContext in GetCctorContextMapping())
+            {
+                // If we generated a static constructor and the owning type, this might be something
+                // that gets fed to RuntimeHelpers.RunClassConstructor. RunClassConstructor
+                // also works on reflection blocked types and there is a possibility that we
+                // wouldn't have generated the cctor otherwise.
+                //
+                // This is a heuristic and we'll possibly root more cctor contexts than
+                // strictly necessary, but it's not worth introducing a new node type
+                // in the compiler just so we can propagate this knowledge from dataflow analysis
+                // (that detects RunClassConstructor usage) and this spot.
+                if (!TypeGeneratesEEType(cctorContext.Type))
+                    continue;
+
+                rootedCctorContexts.Add(cctorContext.Type);
+            }
+
             return new AnalysisBasedMetadataManager(
                 _typeSystemContext, _blockingPolicy, _resourceBlockingPolicy, _metadataLogFile, _stackTraceEmissionPolicy, _dynamicInvokeThunkGenerationPolicy,
                 _modulesWithMetadata, reflectableTypes.ToEnumerable(), reflectableMethods.ToEnumerable(),
-                reflectableFields.ToEnumerable(), _customAttributesWithMetadata);
+                reflectableFields.ToEnumerable(), _customAttributesWithMetadata, rootedCctorContexts);
         }
 
         private struct ReflectableEntityBuilder<T>

--- a/src/tests/nativeaot/SmokeTests/Reflection/Reflection.cs
+++ b/src/tests/nativeaot/SmokeTests/Reflection/Reflection.cs
@@ -27,6 +27,7 @@ internal class ReflectionTest
         //
         // Tests for dependency graph in the compiler
         //
+        TestRunClassConstructor.Run();
 #if !OPTIMIZED_MODE_WITHOUT_SCANNER
         TestContainment.Run();
         TestInterfaceMethod.Run();
@@ -1651,6 +1652,26 @@ internal class ReflectionTest
                 throw new Exception();
 
             if (typeof(TypeWithCodelessMethods).GetMethods(BindingFlags.Public | BindingFlags.Static).Length != 1)
+                throw new Exception();
+        }
+    }
+
+    class TestRunClassConstructor
+    {
+        static class TypeWithNoStaticFieldsButACCtor
+        {
+            static TypeWithNoStaticFieldsButACCtor()
+            {
+                s_cctorRan = true;
+            }
+        }
+
+        private static bool s_cctorRan;
+
+        public static void Run()
+        {
+            RuntimeHelpers.RunClassConstructor(typeof(TypeWithNoStaticFieldsButACCtor).TypeHandle);
+            if (!s_cctorRan)
                 throw new Exception();
         }
     }

